Makefile

```diff
@@ -207,5 +249,65 @@ generate:
 
 lint:
 
-.PHONY: build test install extras clean generate
+check-test-dep:
+	@for TEST_DEP in docker cpp git ; do \
+		if ! which $$TEST_DEP > /dev/null ; then \
+			echo "$$TEST_DEP is required for unit test execution, please install it."; \
+			exit 1; \
+		fi \
+	done
+
+$(TEST_BUILD_TARGETS): test-build-%: check-test-dep
+	@# Skip image build in case that image with recent code changes exists
+	@if [ "`docker images -q $(UT_IMAGE):$*_$(GIT_HASH)`" = "" ] ; then \
+		echo Build unit test image for $*; \
+		cpp -o docker/unit-tests/Dockerfile.$* docker/unit-tests/Dockerfile.$*.in; \
+		docker build . \
+			-t $(UT_IMAGE):$* \
+			-t $(UT_IMAGE):$*_$(GIT_HASH) \
+			--build-arg UT_GO_VERSION=$(UT_GO_VERSION) \
+			-f docker/unit-tests/Dockerfile.$*; \
+	fi
+
+test-build: test-build-$(UT_OS)
+
+test-build-all: $(TEST_BUILD_TARGETS)
+
+$(TEST_TARGETS): test-%: test-build-%
+	@echo Run unit tests at $*
+	@$(SUDO) docker run --rm --privileged $(UT_IMAGE):$*_$(GIT_HASH) bash -c \
+		'UT_LIST=`find . -name "*_test.go" -not -path "./vendor/*" -exec dirname \{\} \; | sort -u`; \
+		for UT_DIR in $$UT_LIST ; do \
+			cd $${UT_DIR}; \
+			go test -v || exit 1; \
+			cd -; \
+		done'
+
+test: install generate test-$(UT_OS)
+
+test-all: $(TEST_TARGETS)
+
+$(COVERAGE_TARGETS): coverage-%: test-build-%
+	@echo Calculate code coverage at $*
+	@$(SUDO) docker run --rm --privileged $(UT_IMAGE):$* bash -c "go-carpet -summary | sort"
+
+coverage: coverage-$(UT_OS)
+
+coverage-all: $(COVERAGE_TARGETS)
+
+test-clean:
+	@if which docker > /dev/null ; then \
+		echo Remove unit test container images with name $(UT_IMAGE); \
+		for IMAGE in `docker images -a -q --filter=reference=$(UT_IMAGE) | sort -u`; do \
+			docker rmi -f $$IMAGE; \
+		done; \
+		echo Remove generated dockerfiles; \
+		for IMAGEOS in $(UT_OS_ALL) ; do \
+			rm -f docker/unit-tests/Dockerfile.$$IMAGEOS; \
+		done; \
```

Comment:
Is there a clever way to add these files to the .gitignore?

Reply:
negating patterns, done in the new commit

Reply:
To pull in your changes, I reset my sandbox and pulled in each commit one by one using:
```
git reset --hard 673eeb0
git pull --no-edit upstream pull/46/head
git pull --no-edit upstream pull/47/head
:
```
PR #56 had a merge conflict on .gitignore. Easy enough to fix manually, but may have an issue when we go to merge these.

Reply:
Once we have this one #47 merged, I'll rebase #56 manually so that there are no conflicts.

Reply:
Now merged.
@przemeklal can we update #56 ?